### PR TITLE
Feat: Added support for cloud_id and cloud_auth

### PR DESCRIPTION
**Pull request by kares**
should be fairly straight-forward: 
- **cloud_id** is a substitute for the *hosts* setting (one can not use both)
- **cloud_auth** is an *user/password* alternative

resolves: https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/893

~~TODO documentation for new options~~

### CODE REVIEW

File: lib/logstash/outputs/elasticsearch.rb [hunk 1]
```diff
@@ -122,6 +122,11 @@ class LogStash::Outputs::ElasticSearch < LogStash::Outputs::Base
   # Password to authenticate to a secure Elasticsearch cluster
   config :password, :validate => :password
 
+  # Cloud authentication string ("<username>:<password>" format) is an alternative for the `user`/`password` configuration.
+  #
+  # For mode details, check out the https://www.elastic.co/guide/en/logstash/current/connecting-to-cloud.html#_cloud_auth[cloud documentation]
+  config :cloud_auth, :validate => :string
```
Comment (robbavey) [inline lib/logstash/outputs/elasticsearch.rb]: This should be `:validate => :password` as it contains password information

```suggestion
  config :cloud_auth, :validate => :password
```

File: lib/logstash/outputs/elasticsearch/common.rb [hunk 2]
```diff
@@ -115,6 +117,64 @@ def setup_hosts
...
+      end
+      @user, @password = parse_user_password_from_cloud_auth(@cloud_auth)
+      params['user'], params['password'] = @user, @password
+    end
+
+    def parse_host_uri_from_cloud_id(cloud_id)
+      begin # might not be available on older LS
+        require 'logstash/util/cloud_setting_id'
+      rescue LoadError
+        raise LogStash::ConfigurationError, 'The cloud.id setting is not supported by LogStash, please set hosts instead.'
```
Comment (robbavey) [inline lib/logstash/outputs/elasticsearch/common.rb]: Nits: `cloud.id` should read `cloud_id` to reflect configuration property name
          Capitalization of "LogStash" should read Logstash
          Maybe state that the setting is not supported by this version of Logstash, as later versions do support it.

File: lib/logstash/outputs/elasticsearch/common.rb [hunk 3]
```diff
@@ -115,6 +117,64 @@ def setup_hosts
...
+      cloud_uri = "#{cloud_id.elasticsearch_scheme}://#{cloud_id.elasticsearch_host}"
+      LogStash::Util::SafeURI.new(cloud_uri)
+    end
+    private :parse_host_uri_from_cloud_id
+
+    def parse_user_password_from_cloud_auth(cloud_auth)
+      begin # might not be available on older LS
+        require 'logstash/util/cloud_setting_auth'
+      rescue LoadError
+        raise LogStash::ConfigurationError, 'The cloud.auth setting is not supported by LogStash, please set user/password instead.'
```
Comment (robbavey) [inline lib/logstash/outputs/elasticsearch/common.rb]: Nits: cloud.id should read cloud_id to reflect configuration property name
Capitalization of "LogStash" should read Logstash
Maybe state that the setting is not supported by this version of Logstash, as later versions do support it.

**Comment by kares** [comment]
circling back here, as this was promised to eventually land in next LS release ...
did try generating docs locally and seemed fine, 
Rob's docs suggestions are still relevant but those won't be addressed here.

**Pull request merged**